src/server/scene/basic_idle_hub.cpp

```diff
@@ -0,0 +1,265 @@
+    poke();
+}
+
+ms::BasicIdleHub::~BasicIdleHub()
+{
+    alarm->cancel();
+}
+
+void ms::BasicIdleHub::poke()
+{
+    std::lock_guard<std::mutex> lock{mutex};
+    poke_time = clock->now();
+    schedule_alarm(lock, poke_time);
+    if (!idle_multiplexers.empty())
+    {
+        auto const idle = std::move(idle_multiplexers);
+        idle_multiplexers.clear();
+        for (auto const& multiplexer : idle)
+        {
+            multiplexer->active();
+        }
+    }
+}
+
+void ms::BasicIdleHub::register_interest(
+    std::weak_ptr<IdleStateObserver> const& observer,
+    time::Duration timeout)
+{
+    register_interest(observer, system_executor, timeout);
+}
+
+void ms::BasicIdleHub::register_interest(
+    std::weak_ptr<IdleStateObserver> const& observer,
+    Executor& executor,
+    time::Duration timeout)
+{
+    auto const shared_observer = observer.lock();
+    if (!shared_observer)
+    {
+        return;
+    }
+
+    std::lock_guard<std::mutex> lock{mutex};
+    auto const iter = timeouts.find(timeout);
+    if (iter == timeouts.end())
+    {
+        auto const multiplexer = std::make_shared<Multiplexer>();
+        multiplexer->register_interest(observer, executor);
+        timeouts.insert({timeout, multiplexer});
+        first_timeout = timeouts.begin()->first;
+        auto const current_time = clock->now() - poke_time;
+        if (!alarm_timeout || alarm_timeout.value() > timeout)
+        {
+            // The alarm will not be fired before we hit our timeout
+            if (current_time < timeout)
+            {
+                // The alarm will be fired after we hit our timout,
+                // but we have not hit it yet so the alarm needs to be moved up
+                alarm_timeout = timeout;
+                alarm->reschedule_in(std::chrono::duration_cast<std::chrono::milliseconds>(timeout - current_time));
+            }
+            else
+            {
+                // This timeout has already been passed
+                idle_multiplexers.push_back(multiplexer);
+            }
+        }
+    }
+    else
+    {
+        iter->second->register_interest(observer, executor);
+    }
+
+    {
+        std::lock_guard<std::recursive_mutex> lock{pending_registrations->mutex};
+        pending_registrations->observers.insert(shared_observer.get());
+    }
+
+    auto const is_active = (alarm_timeout && alarm_timeout.value() <= timeout);
+    executor.spawn([pending_registrations=pending_registrations, shared_observer, is_active]()
+        {
+            std::lock_guard<std::recursive_mutex> lock{pending_registrations->mutex};
+            auto const removed = pending_registrations->observers.erase(shared_observer.get());
+            if (removed)
+            {
+                if (is_active)
+                {
+                    shared_observer->active();
+                }
+                else
+                {
+                    shared_observer->idle();
+                }
+            }
+        });
```

Comment:
Oh! This code is here to make `unregister_interest` guarantee that no further calls will be made to the observer. Urgh, that's gnarly.

I've not quite convinced myself that it's correct yet. I feel like it should *probably* be replaced by adding a way for `ObserverMultiplexer::register_interest` to send an initial state, but I'm not quite sure how that would look. If it stays, it could do with a comment saying what it's for :)